byron-adapter/src/exec/Main.hs

```diff
@@ -0,0 +1,313 @@
+      . Lens.to fromIntegral
+
+
+-- | Use cardano-sl library stuff to get all of the necessary configuration
+-- options. There's a bunch of stuff that we don't actually need but that's OK.
+main :: IO ()
+main = withCompileInfo $ do
+  -- withCompileInfo forced upon us by 'getSimpleNodeOptions'. It uses it to
+  -- print the git revision...
+  -- Take all of the arguments from a cardano-sl node. Hijack db-path and
+  -- reinterpret it as the immutable DB path.
+  SimpleNodeArgs cArgs nArgs <- getSimpleNodeOptions
+  let confOpts = configurationOptions (commonArgs cArgs)
+      -- Take the log config file from the common args, and default it.
+      loggerConfigFile = fromMaybe "logging.yaml" (logConfig (commonArgs cArgs))
+  loggerConfig <- parseLoggerConfig loggerConfigFile
+  lh <- setupLogging' "byron-adapter" loggerConfig
+  -- Should be able to get a trace that we can give to the diffusion layer.
+  trace <- getTrace lh
+  withConfigurations Nothing Nothing False confOpts $ \genesisConfig _ _ _ -> do
+    (nodeParams, Just sscParams) <- getNodeParams
+      "byron-adapter"
+      cArgs
+      nArgs
+      (configGeneratedSecrets genesisConfig)
+    let genesisBlock = genesisBlock0 (configProtocolMagic genesisConfig)
+                                     (configGenesisHash genesisConfig)
+                                     (genesisLeaders genesisConfig)
+        -- Here's all of the config needed for the Byron proxy server itself.
+        bpc :: ByronProxyConfig
+        bpc = ByronProxyConfig
+          { bpcAdoptedBVData = configBlockVersionData genesisConfig
+            -- ^ Hopefully that never needs to change.
+          , bpcNetworkConfig = npNetworkConfig nodeParams
+          , bpcDiffusionConfig = FullDiffusionConfiguration
+              { fdcProtocolMagic = configProtocolMagic genesisConfig
+              , fdcProtocolConstants = configProtocolConstants genesisConfig
+              , fdcRecoveryHeadersMessage = recoveryHeadersMessage
+              , fdcLastKnownBlockVersion = lastKnownBlockVersion updateConfiguration
+              , fdcConvEstablishTimeout = networkConnectionTimeout
+              -- The iohk-monitoring-framework took the `Trace` type and renamed
+              -- it `Cardano.BM.BaseTrace.BaseTrace`. Then `Trace` is defined
+              -- to be `(TraceContext, BaseTrace m NamedLogItem)` where
+              -- `TraceContext` contains configuration stuff and a mysterious
+              -- `IO ()`.
+              --
+              -- The challenge: how to make the original `Trace` from this
+              -- thing, to give to the diffusion layer, in such a way that it
+              -- ehaves as expected? What is the `TraceContext` for? Is it
+              -- needed? It seems all wrong: configuration for the `Trace`
+              -- should be kept _in its closure_ if need be rather than shown
+              -- up-front, and with a mention of `IO` to boot. Very poor
+              -- abstraction.
```

Comment:
@avieth have you communicated this with `iohk-monitoring-framework` guys?

Reply:
I talked to Andreas about it and he claimed there's a need to have some context available for querying it.

Reply:
We need to be able to query and change configuration for `Trace`. For example we want to change minimum severity filter (message with severity less than this will not be logged) at runtime.

But operations on `Trace` cannot return anything but `()`

Reply:
Does that really mean we must query and change the `Trace` values themselves? Shouldn't this be done at the switchboard? The `Trace`s can just throw everything they get at the switchboard, and then it can filter etc.